### ISSUE TITLE
[ Feature Request ] -  Higher bit depth file export

### ISSUE THREAD

**Issue by rethink-studios** [comfyanonymous/ComfyUI#572]
Hey hey,
Is it possible to add export of PNG16, DNG or EXR?
Memory intensive, but would be huge!

**Comment by comfyanonymous** [comment]
Yes, the image data being passed around is RGB values is in float32. The only issue is that I'm not sure if PIL supports saving in higher precision PNG and I don't want to pull in another dependency just for this. 

Should be easy to make a custom node though.

**Comment by WASasquatch** [comment]
The diffusion process may use fload32, but all the source data is 8bit images, so you get no benefit from using fp32 as far as color depth. Just wasted time. The results of the images are within 8bit range minus some slight discrepancy. There was some studies on diffusers that showed fp32 did nothing as far as a difference with fp16, and was just a waste of time and resources. Even then, with only 8bit source data, even fp16 is a waste.

**Comment by WASasquatch** [comment]
PS little more streamlined code usage. It'll take care of the ranging itself based on the file format type.

```python
tiff.imwrite("half-float.tiff", image.cpu().numpy(), dtype=tiff.float16)
tiff.imwrite("full-float.tiff", image.cpu().numpy(), dtype=tiff.float32)
```

**Comment by rethink-studios** [comment]
Last year, while playing with Deforum and wanting AUTO to have access to EXR export, I connected with a developer (Pabla) and they were able to put together a solution, who like @WASasquatch, had issue with how the file saved, and I'm wondering if this helps:

https://stackoverflow.com/questions/42406338/why-cv2-imwrite-changes-the-color-of-pics

Here's Pabla's git, and I know we're not to copy/paste code, however my hope is to shine a light on the subject and help find solutions for us CGI/Print peeps.

https://github.com/pablx-ts/deforum-stable-diffusion


@WASasquatch, for my own education, even is the input data was 8bit, once we're in latent space, and we apply a blur or latent scale to the data, wouldn't the information, at that point, especially if the range of the data has been expanded thru the blur, wouldn't we see SOME benefit? 

Back in the day, I would take multiple bracketed exposures to create an HDR, and I'm wondering if we're getting to the point where we can synthesize f-stops in app. Thoughts?

**Comment by WASasquatch** [comment]
> Last year, while playing with Deforum and wanting AUTO to have access to EXR export, I connected with a developer (Pabla) and they were able to put together a solution, who like @WASasquatch, had issue with how the file saved, and I'm wondering if this helps:
> 
> https://stackoverflow.com/questions/42406338/why-cv2-imwrite-changes-the-color-of-pics
> 
> Here's Pabla's git, and I know we're not to copy/paste code, however my hope is to shine a light on the subject and help find solutions for us CGI/Print peeps.
> 
> https://github.com/pablx-ts/deforum-stable-diffusion
> 
> @WASasquatch, for my own education, even is the input data was 8bit, once we're in latent space, and we apply a blur or latent scale to the data, wouldn't the information, at that point, especially if the range of the data has been expanded thru the blur, wouldn't we see SOME benefit?
> 
> Back in the day, I would take multiple bracketed exposures to create an HDR, and I'm wondering if we're getting to the point where we can synthesize f-stops in app. Thoughts?

A blur won't extend the range of the images actual color,  but just blur the image which sure would have less banding, but wouldn't be truly 16-bit or 32-bit. If you just scaled 8bit data into 16-bit / 32-bit ranges you'll just end up with a black or white image, maybe some faint color burning, cause the ranges are all scaled up, so it's likely blacks and whites will be way out of range.

**Comment by rethink-studios** [comment]
@WASasquatch, I'm wondering if color can be remapped, based on a specified envelope or LUT, while we're in float, that pushes color depth into the high dynamic range space (with a remap node?), before saving the file?

**Comment by WASasquatch** [comment]
> @WASasquatch, I'm wondering if color can be remapped, based on a specified envelope or LUT, while we're in float, that pushes color depth into the high dynamic range space (with a remap node?), before saving the file?

That sounds like probably the right direction. I'll be honest though, I'm no good with tensor/numpy manipulation though.

**Comment by comfyanonymous** [comment]
The problem with generating high dynamic range images with stable diffusion is that the VAE normalizes the image when it converts it to pixel space.

**Comment by WASasquatch** [comment]
> The problem with generating high dynamic range images with stable diffusion is that the VAE normalizes the image when it converts it to pixel space.

Not to mention it's variance with the normal color space as is. Gradients getting blotchy, color loss or strange gain, etc.

**Comment by Foul-Tarnished** [comment]
Would it be useful to save as PNG16 if I'm going to run the image through a ESRGAN-like model later (in chainner) ?
Would it have more infos to upscale more accurately ?

Also if I'm going to print it later, would it print better ?

**Comment by WASasquatch** [comment]
> Would it be useful to save as PNG16 if I'm going to run the image through a ESRGAN-like model later (in chainner) ? Would it have more infos to upscale more accurately ?
> 
> Also if I'm going to print it later, would it print better ?

No. These models are also trained on, and working with 8bit images. There are actually scripts that handle this prior to using ESRGAN cause it isn't supported, such as preserving the alpha as a sep file and adding it back, or converting to 8bit.

**Comment by PeteMatterfield** [comment]
Bringing this one up again. The conversation has been framed entirely within the realm of AI inferencing, but ComfyUI isn't just for AI-related processing; It's excellent for everything else and is becoming more abstract. I use it as an interface to python.

It's becoming an indispensable tool for image and 3D automation and pipeline development where you want to use custom tools, or opencv/pil, or anything else written in python. Much of image,video, and 3D has to do with high bit depth processing, and it's totally and completely worthwhile for the Load Image and Save Image nodes to at least support 16bit.

Sure, I can spin up my own nodes, but it's a bit awkward for a suite of image processing software to only support 8 bit precision out of the box.
